packages/teleport/src/Discover/Shared/SetupAccess/useUserTraits.ts

```diff
@@ -0,0 +1,361 @@
+/**
+ * Copyright 2022 Gravitational, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useState, useEffect } from 'react';
+import useAttempt from 'shared/hooks/useAttemptNext';
+
+import useTeleport from 'teleport/useTeleport';
+import { Option } from 'teleport/Discover/Shared/SelectCreatable';
+
+import { ResourceKind } from '../ResourceKind';
+
+import type { User, UserTraits } from 'teleport/services/user';
+import type { DbMeta, KubeMeta, NodeMeta } from 'teleport/Discover/useDiscover';
+import type { AgentStepProps } from '../../types';
+
+// useUserTraits handles:
+//  - retrieving the latest user (for the dynamic traits) from the backend
+//  - extracting traits into static traits (role-defined) and dynamic traits (user-defined)
+//  - updating user in the backend with the latest dynamic traits
+//  - updating the dynamic traits for our in-memory resource meta object
+//  - provides utility function that makes data objects (type Option) for react-select component
+export function useUserTraits(props: AgentStepProps) {
+  const ctx = useTeleport();
+
+  const [user, setUser] = useState<User>();
+  const { attempt, run, setAttempt, handleError } = useAttempt('processing');
+
+  const isSsoUser = ctx.storeUser.state.authType === 'sso';
+  const canEditUser = ctx.storeUser.getUserAccess().edit;
+  const dynamicTraits = initUserTraits(user);
+
+  // Filter out static traits from the resource that we
+  // queried in a prior step where we discovered the newly connected resource.
+  // The resource itself contains traits that define both
+  // dynamic (user-defined) and static (role-defined) traits.
+  let meta = props.agentMeta;
+  let staticTraits = initUserTraits();
+  switch (props.selectedResourceKind) {
+    case ResourceKind.Kubernetes:
+      const kube = (meta as KubeMeta).kube;
+      staticTraits.kubeUsers = kube.users.filter(
+        v => !dynamicTraits.kubeUsers.includes(v)
+      );
```

Comment:
Can be generalized into difference function (https://proofwiki.org/wiki/Definition:Set_Difference):
```suggestion
const except = (left: string[], right: string[]) =>
  left.filter(l => !right.includes(l));
```